File: bindings/typescript/packages/aster/src/interceptors/rate-limit.ts
/**
 * Token-bucket rate limit interceptor.
 *
 * Enforces rate limits at global, per-service, per-method, and per-peer granularity.
 */

import type { Interceptor } from './base.js';
import { CallContext } from './base.js';
import { RpcError, StatusCode } from '../status.js';

class TokenBucket {
  private tokens: number;
  private lastRefill: number;

  constructor(
    private readonly rps: number,
    private readonly maxBurst: number,
  ) {
    this.tokens = maxBurst;
    this.lastRefill = performance.now();
  }

  tryConsume(): boolean {
    this.refill();
    if (this.tokens >= 1) {
      this.tokens -= 1;
      return true;
    }
    return false;
  }

  private refill(): void {
    const now = performance.now();
    const elapsed = (now - this.lastRefill) / 1000;
    this.tokens = Math.min(this.maxBurst, this.tokens + elapsed * this.rps);
    this.lastRefill = now;
  }
}

export interface RateLimitOptions {
  globalRps?: number;
  perServiceRps?: number;
  perMethodRps?: number;
  perPeerRps?: number;
}

export class RateLimitInterceptor implements Interceptor {
  private global: TokenBucket | undefined;
  private perService = new Map<string, TokenBucket>();
  private perMethod = new Map<string, TokenBucket>();
  private perPeer = new Map<string, TokenBucket>();
  private opts: Required<RateLimitOptions>;

  constructor(opts: RateLimitOptions = {}) {
    this.opts = {
      globalRps: opts.globalRps ?? 0,
      perServiceRps: opts.perServiceRps ?? 0,
      perMethodRps: opts.perMethodRps ?? 0,
      perPeerRps: opts.perPeerRps ?? 0,
    };
    if (this.opts.globalRps > 0) {
      this.global = new TokenBucket(this.opts.globalRps, this.opts.globalRps * 2);
    }
  }

  async onRequest(ctx: CallContext, request: unknown): Promise<unknown> {
    if (this.global && !this.global.tryConsume()) {
      throw new RpcError(StatusCode.RESOURCE_EXHAUSTED, 'global rate limit exceeded');
    }

    if (this.opts.perServiceRps > 0) {
      const bucket = this.getOrCreate(this.perService, ctx.service, this.opts.perServiceRps);
      if (!bucket.tryConsume()) {
        throw new RpcError(StatusCode.RESOURCE_EXHAUSTED, `rate limit exceeded for service ${ctx.service}`);
      }
    }

    if (this.opts.perMethodRps > 0) {
      const key = `${ctx.service}/${ctx.method}`;
      const bucket = this.getOrCreate(this.perMethod, key, this.opts.perMethodRps);
      if (!bucket.tryConsume()) {
        throw new RpcError(StatusCode.RESOURCE_EXHAUSTED, `rate limit exceeded for ${key}`);
      }
    }

    if (this.opts.perPeerRps > 0 && ctx.peer) {
      const bucket = this.getOrCreate(this.perPeer, ctx.peer, this.opts.perPeerRps);
      if (!bucket.tryConsume()) {
        throw new RpcError(StatusCode.RESOURCE_EXHAUSTED, `rate limit exceeded for peer ${ctx.peer}`);
      }
    }

    return request;
  }

  private getOrCreate(map: Map<string, TokenBucket>, key: string, rps: number): TokenBucket {
    let bucket = map.get(key);
    if (!bucket) {
      bucket = new TokenBucket(rps, rps * 2);
      map.set(key, bucket);
    }
    return bucket;
  }
}

File: bindings/python/aster/interceptors/rate_limit.py
"""
aster.interceptors.rate_limit — Token bucket rate limiter for RPC calls.

Limits request rate per service, per method, or per peer. Rejects requests
that exceed the limit with RESOURCE_EXHAUSTED status.

Usage::

    from aster.interceptors import RateLimitInterceptor

    # 100 requests/second globally
    server = AsterServer(
        services=[...],
        interceptors=[RateLimitInterceptor(rate=100)],
    )

    # Per-service limits
    server = AsterServer(
        services=[...],
        interceptors=[RateLimitInterceptor(rate=100, per="service")],
    )

    # Per-peer limits (useful for multi-tenant)
    server = AsterServer(
        services=[...],
        interceptors=[RateLimitInterceptor(rate=50, per="peer")],
    )
"""

from __future__ import annotations

import time
from typing import Any

from aster.interceptors.base import CallContext, Interceptor
from aster.status import RpcError, StatusCode


class _TokenBucket:
    """Simple token bucket rate limiter."""

    __slots__ = ("_rate", "_capacity", "_tokens", "_last_refill")

    def __init__(self, rate: float, burst: float | None = None) -> None:
        self._rate = rate  # tokens per second
        self._capacity = burst or rate  # max burst
        self._tokens = self._capacity
        self._last_refill = time.monotonic()

    def try_acquire(self) -> bool:
        """Try to consume one token. Returns True if allowed."""
        now = time.monotonic()
        elapsed = now - self._last_refill
        self._tokens = min(self._capacity, self._tokens + elapsed * self._rate)
        self._last_refill = now

        if self._tokens >= 1.0:
            self._tokens -= 1.0
            return True
        return False


class RateLimitInterceptor(Interceptor):
    """Token bucket rate limiter interceptor.

    Args:
        rate: Maximum requests per second.
        burst: Maximum burst size (defaults to rate).
        per: Granularity: ``"global"``, ``"service"``, ``"method"``, or ``"peer"``.
    """

    def __init__(
        self,
        rate: float = 100.0,
        burst: float | None = None,
        per: str = "global",
    ) -> None:
        self._rate = rate
        self._burst = burst
        self._per = per
        self._buckets: dict[str, _TokenBucket] = {}
        self._global_bucket = _TokenBucket(rate, burst)

    def _get_bucket(self, ctx: CallContext) -> _TokenBucket:
        if self._per == "global":
            return self._global_bucket

        if self._per == "service":
            key = ctx.service
        elif self._per == "method":
            key = f"{ctx.service}.{ctx.method}"
        elif self._per == "peer":
            key = ctx.peer or "unknown"
        else:
            return self._global_bucket

        if key not in self._buckets:
            self._buckets[key] = _TokenBucket(self._rate, self._burst)
        return self._buckets[key]

    async def on_request(self, ctx: CallContext, request: object) -> object:
        bucket = self._get_bucket(ctx)
        if not bucket.try_acquire():
            raise RpcError(
                StatusCode.RESOURCE_EXHAUSTED,
                f"Rate limit exceeded ({self._rate}/s per {self._per})",
            )
        return request
